### PR TITLE
[v6r20] UserProxy contextManager, StorageManager, JobScheduling fix

### LINKED ISSUE #3764: [v6r20] Bug in StorageManager checkFilesToStage
Every  once in a while the JobScheduling executor has problems to get the StorageMetadata for some inputs to some jobs. I finally tracked this down to these lines (putting a traceback printout in the writeProxyToFile function):

https://github.com/DIRACGrid/DIRAC/blob/894421bccf2a23300f5fdab6611056006c7fc694/StorageManagementSystem/Client/StorageManagerClient.py#L142-L143
...
https://github.com/DIRACGrid/DIRAC/blob/894421bccf2a23300f5fdab6611056006c7fc694/StorageManagementSystem/Client/StorageManagerClient.py#L161-L165

This separation of StorageElement with and without userProxy doesn't work, because of the gfal2 context, which caches the proxy location, that is part of the StorageElement.

The StorageElement instance needs to be obtained/created when the userProxy exists, and not before.
Thus we either have to wrap this call in another function that gets decorated with `@executeWithUserProxy`, which isn't very nice, or we need to create a contextManager to run
```python
with userProxy(proxyUserName....) as proxyResult:
  if proxyResult['OK']:
    StorageElement(...).getFileMetadata(...)
```

### PR DESCRIPTION
**Pull request by andresailer**
I realise this is a lot of refactoring changes for a patch, and I want to test this more thoroughly, but I put it here now for discussion.

BEGINRELEASENOTES

*CORE
NEW: New Core.Utilities.Proxy.UserProxy class to be used as a contextManager
FIX: Fix race condition in Core.Utilities.Proxy.executeWithUserProxy, the $X509_USER_PROXY environment variable from one thread could leak to another, fixes #3764

*SMS:
FIX: Fix setting of the user proxy for StorageElement.getFileMetadata calls, fixes #3764

*WMS
FIX: The solution for bug #3764 fixes a problem with the JobScheduling executor, where files could end up in the checking state with the error "Couldn't get storage metadata of some files"

ENDRELEASENOTES

### CODE REVIEW

**Comment by fstagni** [comment]
This `executeWithUserProxy` is used in some other places, do you mind updating them too?

**Comment by andresailer** [comment]
The other uses of executeWithUserProxy that are python files and not import lines
```
$ git grep executeWithUserProxy | grep -v import | grep "\.py:"
ConfigurationSystem/Agent/VOMS2CSAgent.py:  @executeWithUserProxy
ConfigurationSystem/Agent/VOMS2CSAgent.py:  @executeWithUserProxy
ConfigurationSystem/scripts/dirac-admin-voms-sync.py:@executeWithUserProxy
TransformationSystem/Agent/TransformationCleaningAgent.py:          executeWithUserProxy(self._executeClean)(transDict,
TransformationSystem/Agent/TransformationCleaningAgent.py:          executeWithUserProxy(self._executeRemoval)(transDict,
TransformationSystem/Agent/TransformationCleaningAgent.py:          executeWithUserProxy(self._executeArchive)(transDict,
WorkloadManagementSystem/Executor/InputData.py:  @executeWithUserProxy
WorkloadManagementSystem/Service/WMSUtilities.py:@executeWithUserProxy
```
The only place were executeWithUserProxy is used as it was here is in the TransformationCleaningAgent, were just a single functions are called. There is no benefit in changing that to `with UserProxy...`.
For the other cases changes would surely be much more intrusive

**Comment by chaen** [comment]
Seems OK to me

File: Core/Utilities/Proxy.py [hunk 1]
```diff
@@ -183,3 +172,98 @@ def wrapped_fcn( *args, **kwargs ):
...
+    if not result['OK']:
+      return result
+
+    executionLock = LockRing().getLock('_UseUserProxy_', recursive=True) if executionLockFlag else None
+    if executionLockFlag:
+      executionLock.acquire()
+
+    proxyFile = result['Value']
+    originalUserProxy = os.environ.get('X509_USER_PROXY')
+    os.environ['X509_USER_PROXY'] = proxyFile
```
Comment (fstagni) [inline Core/Utilities/Proxy.py]: Minor, but you can do this switch in just one line:
```
os.environ['X509_USER_PROXY''], originalUserProxy = result['Value'], os.environ.get('X509_USER_PROXY')
```

Comment (andresailer) [reply Core/Utilities/Proxy.py]: Done

**Comment by andresailer** [comment]
I hot fixed this into the production servers, so seems to be fine.

**Pull request merged**